### ISSUE TITLE
segfault following "Supplied genesis block does not match chain data stored"

### ISSUE THREAD

**Issue by macfarla** [hyperledger/besu#6279]
Running on my local machine. Previously this node was syncing linea goerli. I switched the config to point to a different genesis file and startup fails which is expected. However - segfault - happens reliably, running via IDE or on command line.  

```
####################################################################################################
#                                                                                                  #
# Besu version 23.10.4-dev-fb7d8854                                                                #
#                                                                                                  #
# Configuration:                                                                                   #
# Network: Custom genesis file specified                                                           #
# Network Id: 648529                                                                               #
# Data storage: Forest                                                                             #
# Sync mode: Snap                                                                                  #
# RPC HTTP APIs: ADMIN,ETH,NET,WEB3,PERM,DEBUG,MINER,EEA,PRIV,TXPOOL                               #
# RPC HTTP port: 8565                                                                              #
# Using LAYERED transaction pool implementation                                                    #
# Using STACKED worldstate update mode                                                             #
#                                                                                                  #
# Host:                                                                                            #
# Java: openjdk-java-17                                                                            #
# Maximum heap size: 24.00 GB                                                                      #
# OS: osx-x86_64                                                                                   #
# Total memory: 96.00 GB                                                                           #
# CPU cores: 12                                                                                    #
#                                                                                                  #
# Plugins:                                                                                         #
# TOTAL = 0 of 0 plugins successfully loaded                                                       #
# from /Users/sallymacfarlane/workspace/b2/build/install/besu/plugins                              #
#                                                                                                  #
####################################################################################################
```
```
2023-12-11 14:32:27.769+10:00 | main | INFO  | ProtocolScheduleBuilder | Protocol schedule created with milestones: [Petersburg:0, Berlin:37044085]
2023-12-11 14:32:27.850+10:00 | main | INFO  | DatabaseMetadata | Lookup database metadata file in data directory: /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data
2023-12-11 14:32:27.877+10:00 | main | INFO  | RocksDBKeyValueStorageFactory | Existing database detected at /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data. Version 1. Compacting database...
2023-12-11 14:32:27.878+10:00 | main | DEBUG | RocksDBKeyValueStorageFactory | FOREST mode detected, using TransactionDB.
2023-12-11 14:32:29.238+10:00 | main | ERROR | Besu | Failed to start Besu Supplied genesis block does not match chain data stored in /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data.
Please specify a different data directory with --data-path, specify the original genesis file with --genesis-file or supply a testnet/mainnet option with --network.
picocli.CommandLine$ParameterException: Supplied genesis block does not match chain data stored in /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data.
Please specify a different data directory with --data-path, specify the original genesis file with --genesis-file or supply a testnet/mainnet option with --network.
	at org.hyperledger.besu.cli.BesuCommand.run(BesuCommand.java:1479)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:62)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:33)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1625)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1422)
	at org.hyperledger.besu.Besu.main(Besu.java:39)
Caused by: picocli.CommandLine$ExecutionException: Supplied genesis block does not match chain data stored in /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data.
Please specify a different data directory with --data-path, specify the original genesis file with --genesis-file or supply a testnet/mainnet option with --network.
	at org.hyperledger.besu.cli.BesuCommand.buildController(BesuCommand.java:2164)
	at org.hyperledger.besu.cli.BesuCommand.initController(BesuCommand.java:2150)
	at org.hyperledger.besu.cli.BesuCommand.run(BesuCommand.java:1462)
	... 16 more
Caused by: org.hyperledger.besu.util.InvalidConfigurationException: Supplied genesis block does not match chain data stored in /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data.
Please specify a different data directory with --data-path, specify the original genesis file with --genesis-file or supply a testnet/mainnet option with --network.
	at org.hyperledger.besu.ethereum.chain.DefaultBlockchain.setGenesis(DefaultBlockchain.java:785)
	at org.hyperledger.besu.ethereum.chain.DefaultBlockchain.lambda$new$0(DefaultBlockchain.java:106)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at org.hyperledger.besu.ethereum.chain.DefaultBlockchain.<init>(DefaultBlockchain.java:106)
	at org.hyperledger.besu.ethereum.chain.DefaultBlockchain.createMutable(DefaultBlockchain.java:228)
	at org.hyperledger.besu.controller.BesuControllerBuilder.build(BesuControllerBuilder.java:605)
	at org.hyperledger.besu.cli.BesuCommand.buildController(BesuCommand.java:2162)
	... 18 more

To display full help:
besu [COMMAND] --help
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000001a670bc38, pid=27789, tid=45827
#
# JRE version: OpenJDK Runtime Environment Homebrew (17.0.5) (build 17.0.5+0)
# Java VM: OpenJDK 64-Bit Server VM Homebrew (17.0.5+0, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
# C  [librocksdbjni10398288656630895137.jnilib+0x3dbc38]  rocksdb::FullTypedCacheHelperFns<rocksdb::Block_kData, rocksdb::BlockCreateContext>::Create(rocksdb::Slice const&, rocksdb::Cache::CreateContext*, rocksdb::MemoryAllocator*, void**, unsigned long*)+0x7e8
#
# Core dump will be written. Default location: /cores/core.27789
#
# An error report file with more information is saved as:
# /Users/sallymacfarlane/workspace/besu-local-nodes/hs_err_pid27789.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/Homebrew/homebrew-core/issues
```

**Comment by macfarla** [comment]
config file

```
data-path="workdir/besu3/data"
logging="DEBUG"
sync-mode="X_SNAP"
fast-sync-min-peers=1

Xdns-enabled=true

genesis-file="config/besu/genesis.json"
#genesis-file="config/besu/genesis-linea-goerli.json"
node-private-key-file="workdir/besu3/keys/besu3.priv"

miner-enabled=false
miner-coinbase="0x0000000000000000000000000000000000000000"


p2p-host="1.146.127.172"
p2p-port=50505

min-gas-price=0

host-allowlist=["*"]

rpc-http-enabled=true
rpc-http-api=["ADMIN","ETH","NET","WEB3","PERM","DEBUG","MINER","EEA","PRIV","TXPOOL"]
rpc-http-host="0.0.0.0"
rpc-http-port=8565
rpc-http-cors-origins=["*"]

rpc-ws-enabled=true
rpc-ws-api=["ADMIN","ETH","NET","WEB3","PERM","DEBUG","MINER","EEA","PRIV","TXPOOL"]
rpc-ws-host="0.0.0.0"
rpc-ws-port=8566
```

**Comment by macfarla** [comment]
if I add `data-storage-format="BONSAI"` to the config, I still get a segfault but with a different error
```
2023-12-12 07:14:15.424+10:00 | main | INFO  | ProtocolScheduleBuilder | Protocol schedule created with milestones: [Petersburg:0]
2023-12-12 07:14:15.519+10:00 | main | INFO  | DatabaseMetadata | Lookup database metadata file in data directory: /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data
2023-12-12 07:14:15.603+10:00 | main | INFO  | RocksDBKeyValueStorageFactory | Existing database detected at /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data. Version 1. Compacting database...
2023-12-12 07:14:15.614+10:00 | main | DEBUG | RocksDBKeyValueStorageFactory | Using OptimisticTransactionDB.
2023-12-12 07:14:16.488+10:00 | main | ERROR | Besu | Failed to start Besu
picocli.CommandLine$ExecutionException: Column handle not found for segment TRIE_BRANCH_STORAGE
	at org.hyperledger.besu.cli.BesuCommand.buildController(BesuCommand.java:2164)
	at org.hyperledger.besu.cli.BesuCommand.initController(BesuCommand.java:2150)
	at org.hyperledger.besu.cli.BesuCommand.run(BesuCommand.java:1462)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:62)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:33)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1625)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1422)
	at org.hyperledger.besu.Besu.main(Besu.java:39)
Caused by: java.lang.RuntimeException: Column handle not found for segment TRIE_BRANCH_STORAGE
	at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage.safeColumnHandle(RocksDBColumnarKeyValueStorage.java:305)
	at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage.get(RocksDBColumnarKeyValueStorage.java:316)
	at org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage.deriveFlatDbStrategy(BonsaiWorldStateKeyValueStorage.java:128)
	at org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage.loadFlatDbStrategy(BonsaiWorldStateKeyValueStorage.java:111)
	at org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage.<init>(BonsaiWorldStateKeyValueStorage.java:93)
	at org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStorageProvider.createWorldStateStorage(KeyValueStorageProvider.java:79)
	at org.hyperledger.besu.controller.BesuControllerBuilder.build(BesuControllerBuilder.java:599)
	at org.hyperledger.besu.cli.BesuCommand.buildController(BesuCommand.java:2162)
	... 18 more
picocli.CommandLine$ParameterException: Column handle not found for segment TRIE_BRANCH_STORAGE
	at org.hyperledger.besu.cli.BesuCommand.run(BesuCommand.java:1479)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:62)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:33)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1625)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1422)
	at org.hyperledger.besu.Besu.main(Besu.java:39)
Caused by: picocli.CommandLine$ExecutionException: Column handle not found for segment TRIE_BRANCH_STORAGE
	at org.hyperledger.besu.cli.BesuCommand.buildController(BesuCommand.java:2164)
	at org.hyperledger.besu.cli.BesuCommand.initController(BesuCommand.java:2150)
	at org.hyperledger.besu.cli.BesuCommand.run(BesuCommand.java:1462)
	... 16 more
Caused by: java.lang.RuntimeException: Column handle not found for segment TRIE_BRANCH_STORAGE
	at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage.safeColumnHandle(RocksDBColumnarKeyValueStorage.java:305)
	at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage.get(RocksDBColumnarKeyValueStorage.java:316)
	at org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage.deriveFlatDbStrategy(BonsaiWorldStateKeyValueStorage.java:128)
	at org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage.loadFlatDbStrategy(BonsaiWorldStateKeyValueStorage.java:111)
	at org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage.<init>(BonsaiWorldStateKeyValueStorage.java:93)
	at org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStorageProvider.createWorldStateStorage(KeyValueStorageProvider.java:79)
	at org.hyperledger.besu.controller.BesuControllerBuilder.build(BesuControllerBuilder.java:599)
	at org.hyperledger.besu.cli.BesuCommand.buildController(BesuCommand.java:2162)
	... 18 more

To display full help:
besu [COMMAND] --help
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000001a7ed1c38, pid=90444, tid=45107
#
# JRE version: OpenJDK Runtime Environment Homebrew (17.0.5) (build 17.0.5+0)
# Java VM: OpenJDK 64-Bit Server VM Homebrew (17.0.5+0, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
# C  [librocksdbjni7928008421506099992.jnilib+0x3dbc38]  rocksdb::FullTypedCacheHelperFns<rocksdb::Block_kData, rocksdb::BlockCreateContext>::Create(rocksdb::Slice const&, rocksdb::Cache::CreateContext*, rocksdb::MemoryAllocator*, void**, unsigned long*)+0x7e8
```

**Comment by garyschulte** [comment]
@macfarla are you still able to reproduce this on main?  I have tried various combinations of forest and bonsai, old devnets, linea, etc - and I haven't been able to reproduce the segfault on a Mac M* machine.  If you can still reproduce it would you attach the genesis file you are using ?

**Comment by macfarla** [comment]
yep I'm still able to reproduce on latest main 658d153f3
This is the genesis the chain was started with 
[genesis-linea-goerli.json](https://github.com/hyperledger/besu/files/13854961/genesis-linea-goerli.json)

I also just tried creating a new db from scratch with that genesis, and get the same error if I comment out the genesis config - so don't think it's anything weird with that particular db

**Comment by macfarla** [comment]
@garyschulte 

I just tried again and no longer get a segfault - did you fix it?? (actually I rebuilt from 658d153 and couldn't reproduce it either) - closing anyhow

```
2024-01-31 12:06:37.765+10:00 | main | ERROR | Besu | Failed to start Besu Supplied genesis block does not match chain data stored in /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data.
Please specify a different data directory with --data-path, specify the original genesis file with --genesis-file or supply a testnet/mainnet option with --network.
picocli.CommandLine$ParameterException: Supplied genesis block does not match chain data stored in /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data.
Please specify a different data directory with --data-path, specify the original genesis file with --genesis-file or supply a testnet/mainnet option with --network.
	at org.hyperledger.besu.cli.BesuCommand.run(BesuCommand.java:1141)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:66)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:33)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1289)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1085)
	at org.hyperledger.besu.Besu.main(Besu.java:39)
Caused by: picocli.CommandLine$ExecutionException: Supplied genesis block does not match chain data stored in /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data.
Please specify a different data directory with --data-path, specify the original genesis file with --genesis-file or supply a testnet/mainnet option with --network.
	at org.hyperledger.besu.cli.BesuCommand.buildController(BesuCommand.java:1820)
	at org.hyperledger.besu.cli.BesuCommand.initController(BesuCommand.java:1806)
	at org.hyperledger.besu.cli.BesuCommand.run(BesuCommand.java:1124)
	... 16 more
Caused by: org.hyperledger.besu.util.InvalidConfigurationException: Supplied genesis block does not match chain data stored in /Users/sallymacfarlane/workspace/besu-local-nodes/workdir/besu3/data.
Please specify a different data directory with --data-path, specify the original genesis file with --genesis-file or supply a testnet/mainnet option with --network.
	at org.hyperledger.besu.ethereum.chain.DefaultBlockchain.setGenesis(DefaultBlockchain.java:785)
	at org.hyperledger.besu.ethereum.chain.DefaultBlockchain.lambda$new$0(DefaultBlockchain.java:106)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at org.hyperledger.besu.ethereum.chain.DefaultBlockchain.<init>(DefaultBlockchain.java:106)
	at org.hyperledger.besu.ethereum.chain.DefaultBlockchain.createMutable(DefaultBlockchain.java:228)
	at org.hyperledger.besu.controller.BesuControllerBuilder.build(BesuControllerBuilder.java:608)
	at org.hyperledger.besu.cli.BesuCommand.buildController(BesuCommand.java:1818)
	... 18 more

To display full help:
besu [COMMAND] --help
```
